### PR TITLE
[IE CLDNN] Plugin-side kernels caching

### PR DESCRIPTION
**Pull request by vladimir-paramuzov**
- Added plugin key to enable kernels caching and added plugin-side mechanism to do it.

### CODE REVIEW

File: inference-engine/src/cldnn_engine/cldnn_config.h [hunk 1]
```diff
@@ -59,6 +60,7 @@ struct Config {
     std::string graph_dumps_dir;
     std::string sources_dumps_dir;
     std::string device_id;
+    std::string kernels_cache_dir;
```
Comment (ilya-lavrenov) [inline inference-engine/src/cldnn_engine/cldnn_config.h:63]: I think we need to take into account that path specified by a user can contain unicode symbols.

**Comment by JulienMaille** [comment]
How is this going to work if we have openCL `cl_cache` folder mechanism already enabled?

**Comment by vladimir-paramuzov** [comment]
> How is this going to work if we have openCL `cl_cache` folder mechanism already enabled?

If `cl_cache` is enabled as well, then: 
 - compiled binaries will be dumped twice
 - if a kernel is found in plugin's cache on load stage, then `cl_cache` is not used for this kernel

File: inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.cpp [hunk 2]
```diff
@@ -28,13 +28,89 @@
...
+        auto res = fread(ret.data(), sizeof(unsigned char), nsize, fp);
+        (void)res;
+        fclose(fp);
+        return ret;
+    }
+
+    return {};
+}
+
+static void saveBinaryToFile(std::string path, const std::vector<unsigned char> buffer) {
```
Comment (ilya-lavrenov) [inline inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.cpp:99]: can we add a test for such case? See `smoke_registerPluginsXMLUnicodePath` for example..

File: inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.cpp [hunk 3]
```diff
@@ -140,23 +242,30 @@ kernels_cache::sorted_code kernels_cache::get_program_source(const kernels_code&
...
+    // Compute hash value for each bucket
+    // Hash calculation might require additional optimizations, but currently execution time of this part is much smaller than loading
+    // of the precompiled binaries or get_undef_jit calls
+    for (auto& c : scode) {
+        program_code& code = c.second;
+        for (size_t i = 0; i < code.source.size(); i++) {
+            std::string full_code;
+            for (auto& ss : code.source[i])
+                full_code += ss;
+            code.hash_values.push_back(std::hash<std::string>()(full_code));
```
Comment (ilya-lavrenov) [inline inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.cpp:267]: BTW, hash is a source code? Maybe we needs some notes about has compute part, e.g. how it's computed

Comment (vladimir-paramuzov) [reply inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.cpp:270]: added a comment that explains what exactly is used for hash computation. Do you think it should be duplicated somewhere in the docs or we don't need such kind of implementation details in the docs?

File: inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.cpp (hunk 2 above)
Comment (vladimir-paramuzov) [reply inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.cpp:99]: added test with unicode characters in path

File: inference-engine/src/cldnn_engine/cldnn_config.h (hunk 1 above)
Comment (vladimir-paramuzov) [reply inference-engine/src/cldnn_engine/cldnn_config.h:63]: added unicode support

**Comment by vladimir-paramuzov** [comment]
@iefode please have a look at the tests and test utils

File: inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.cpp (hunk 3 above)
Comment (ilya-lavrenov) [reply inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.cpp:270]: I think inline comments in code are enough.

File: inference-engine/tests/ie_test_utils/common_test_utils/w_dirent.h [hunk 4]
```diff
@@ -0,0 +1,219 @@
+﻿// Copyright (C) 2018-2020 Intel Corporation
```
Comment (iefode) [inline inference-engine/tests/ie_test_utils/common_test_utils/w_dirent.h:1]: `// Copyright (C) 2020 Intel Corporation`

File: inference-engine/tests/ie_test_utils/common_test_utils/w_dirent.h [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+# define NOMINMAX_UNDEF
+#endif
+
+#if defined(_M_IX86) && !defined(_X86_) && !defined(_AMD64_)
+# define _X86_
+#endif
+
+#if defined(_M_X64) && !defined(_X86_) && !defined(_AMD64_)
+# define _AMD64_
+#endif
```
Comment (ilya-lavrenov) [inline inference-engine/tests/ie_test_utils/common_test_utils/w_dirent.h:25]: please, add https://github.com/openvinotoolkit/openvino/pull/2928/files#diff-efb548b49a1ad8370ef87618ac4a2f06ca421322c2f5f86f2dcce89a0a31a63aR27-R33

**Review by iefode — APPROVED** [review]

**Review by ilya-lavrenov — APPROVED** [review]

File: inference-engine/tests/ie_test_utils/common_test_utils/w_dirent.h (hunk 5 above)
Comment (vladimir-paramuzov) [reply inference-engine/tests/ie_test_utils/common_test_utils/w_dirent.h:25]: Added

File: inference-engine/tests/ie_test_utils/common_test_utils/w_dirent.h (hunk 4 above)
Comment (vladimir-paramuzov) [reply inference-engine/tests/ie_test_utils/common_test_utils/w_dirent.h]: Fixed

**Review by sshlyapn — APPROVED** [review]
LGTM

**Pull request merged**